### PR TITLE
Fix 173.

### PR DESCRIPTION
**Pull request by gregglind**
This features should work:

- setup
- sendTelemetry
- endStudy
- uninstall.

These are delayed:
- log
- dataPermissions (and reacting to it.)

Tests:

- npm run test

### CODE REVIEW

File: test-addon/src/background.js [hunk 1]
```diff
@@ -81,9 +82,9 @@ class StudyLifeCycleHandler {
  * @returns {Promise<void>} Nothing
  */
 async function onEveryExtensionLoad() {
-  new StudyLifeCycleHandler();
-  const studySetup = await getStudySetup();
-  await browser.study.setup(studySetup);
+  // new StudyLifeCycleHandler();
+  // const studySetup = await getStudySetup();
+  // await browser.study.setup(studySetup);
```
Comment (punamdahiya) [inline test-addon/src/background.js]: Line 37 we can comment/remove all console.* as long as we have sufficient logging of add-on install and uninstall in api.js

File: test-addon/src/studySetup.js [hunk 2]
```diff
@@ -87,8 +86,8 @@ const studySetup = {
   // Optional: testing overrides.
   // Set from prefs in getStudySetup
   testing: {
-    variation: null,
-    firstRunTimestamp: null,
+    variation: "feature-active",
+    expired: true,
```
Comment (punamdahiya) [inline test-addon/src/studySetup.js]: Quick question, final version supports testing variation by picking variation name from pref defined as property inside studySetup e.g.

variationOverridePreference: "cloud.services.shield.variation",

Comment (gregglind) [reply test-addon/src/studySetup.js]: Yes.   That should work.

File: test-addon/src/background.js (hunk 1 above)
Comment (punamdahiya) [reply test-addon/src/background.js]: @gregglind  I assume studyInfo.timeUntilExpire is calculated and different from studyInfo.expire defined inside studySetup.js?

File: examples/small-study/src/background.js [hunk 3]
```diff
@@ -98,6 +98,8 @@ class StudyLifeCycleHandler {
         // uninstall the addon?
         break;
     }
+    // actually remove the addon.
+    await browser.study.uninstall();
```
Comment (rhelmer) [inline examples/small-study/src/background.js]: any reason this isn't done in the `default` case above? That will run regardless of reason too right?

File: bin/verifyWeeSchema.js [hunk 4]
```diff
@@ -34,16 +43,42 @@ for (const i in proposed) {
...
+
+${full(tc)}
+          `,
+        );
+        console.error(ajv.errors);
+      }
+    }
+
+    // now, known failures
+    for (const tc of type.failcases || []) {
```
Comment (rhelmer) [inline bin/verifyWeeSchema.js]: This is confusing to me... when would `const tc of type.failcases` be falsy? Could the default of that be set to `[]` instead?

File: test/functional/test-addon.js [hunk 5]
```diff
@@ -87,24 +87,23 @@ describe("Tests verifying that the test add-on works as expected", function() {
...
-              endingClass: "ended-positive",
-            });
-            callback();
-          },
-        );
-      });
+    });*/
 
+    describe.skip("test the library's endStudy() function", function() {
+      // before(async() => {
```
Comment (rhelmer) [inline test/functional/test-addon.js]: Why check in commented-out stuff? I'd just remove it, it'll be in the commit history if you really need to dig it up. Or at least use `/* */` :)

**Comment by gregglind** [comment]
Tests pass, provisional R+ from @rhelmer, merging to develop.

**Pull request merged**